main/src/ca/uwaterloo/flix/language/phase/Enumgen.scala

```diff
@@ -0,0 +1,353 @@
+package ca.uwaterloo.flix.language.phase
+
+import org.objectweb.asm
+import org.objectweb.asm.{ClassWriter, Label}
+import org.objectweb.asm.Opcodes._
+import ca.uwaterloo.flix.language.ast.{Type, _}
+import ca.uwaterloo.flix.util.InternalCompilerException
+
+/**
+  * Created by ramin on 2017-05-22.
+  *
+  * Generates classes and interfaces which will represent enums
+  */
+object Enumgen {
+  import CodegenHelper._
+
+  /**
+    * Generates an interface for TagInterface which includes one method which
+    * returns the name of the enum case which implements this interface
+    * @param name name of the class
+    * @return byte code representation of the class
+    */
+  def compileTagInterface(name: List[String]) : Array[Byte] = {
```

Comment:
Passing a `List[String]` around as the name of the class is not only in my code. It is used everywhere. Also, you have to make this conversion at more places if you don't do it in earlier steps. I'm actually refactoring all the places that we use `List[String]` and I'm using a case class called `QualName(ref: List[String])`. `QualName` comes from `QualifiedName` which is basically address of classes in Java. If you don't like this convention then again we can change it, but take a look at the code which I'll push shortly to see if this is an acceptable solution.